### ISSUE TITLE
Spruce does not seem to be growing in Taiga Hills / acacia not growing in deserts

### ISSUE THREAD

**Issue by drjawa** [Civcraft/RealisticBiomes#8]
https://www.reddit.com/r/Civcraft/comments/2mzr1c/civcraft_changelog_for_20141121/cm92alq

tldr; spruce reportedly not growing in taiga hills. Game reports 8% growth rate, but still stuck at saplings out of hundreds of trees. Similarly, growing acacia in desert seems to have similar issue.

**Comment by staygroovy57** [comment]
@WildWeazel Do you need me to do anything else with regards to testing on this?

**Comment by WildWeazel** [comment]
@staygroovy57 get @drjawa to check out the location, we've had sporadic reports of this but never able to verify

**Comment by ttk2** [comment]
whats the location?

**Comment by staygroovy57** [comment]
As listed in the new issue I had started coords are: -11500 -9250 (79)

For the sake of eliminating confusion, and *since this one has been reopened* what should I do with this new issue I had created?

**Comment by ttk2** [comment]
just close it, we want to keep the issue with the most contex

**Comment by staygroovy57** [comment]
Hello "Six Months ago"... been patiently waiting for this to be looked at. Now is the time!

**Comment by ttk2** [comment]
well someone needs to submit a pull.

**Comment by WildWeazel** [comment]
Still no proof of either not growing correctly. I need a screenshot with debug (F3) enabled and the chat message showing the rate for the item in your hand.

Now someone is saying that the taiga biome is very rare, another reason to grow spruce in ice mountains. No idea how accurate that is.

**Comment by staygroovy57** [comment]
@WildWeazel  Here is the "[**proof**](http://i.imgur.com/zuHYQIa.png)" you requested. 

It should be noted that:

- it clearly says 'TaigaHills' (not 'Taiga', *where spruce grows just fine, thank you very much*)
- spruce used to grow as expected in this area before whatever changes were made around the time Dark Oak and Acacia were finally edited into the RB config. 

Additionally, the Taiga biome comment fro 'someone' about them being rare would seem to arguably come from a person who hasn't traveled the map and explored very much. I beg to differ, there are plenty of them all over the place, but some are called different things. Like for example the one NW of Orion is called something like 'Taiga Plains' or 'Ice Plains' IIRC?

**Comment by WildWeazel** [comment]
Huh, so there it is. That matches what the config says, but their behavior doesn't match the reported rate. I wonder if there is some nuance to vanilla spruce behavior in Taiga Hills specifically? Both of the large spruce types also grow at 8% in Taiga, so those shouldn't affect it.

BTW Ice Plains/Mountains, sometimes referred to as Tundra, are a distinct biome type and configured separately, under our "cold" category but not "cold forest"

**Comment by staygroovy57** [comment]
Spruce used to grow there is all I can say. 

Now on to acacia in desert (only thing is: that problem is at level 235, although I have trees growing fine at other sky farms in other biomes like Jungle, this one doesn't)

**Comment by WildWeazel** [comment]
Since trees are not persisted it's entirely possible that some vanilla rule introduced in an update is preventing them from attempting to grow in certain locations.

**Comment by staygroovy57** [comment]
@WildWeazel  Here is [a link to a screenshot](http://i.imgur.com/JPfvlXi.png) showing the desert area where acacia doesn't appear to be growing.

**Comment by WildWeazel** [comment]
Thanks. Looking at the config, rates should be as follows:

Acacia
Desert, desert hills 8%
Plains 4%
All others 0%

Dark Oak
Swamp 8%
Forest, forest hills 4%
Taiga, taiga hills 2%
All others 0%

I wonder if there are any other exceptions to this. I'd also like to see if we can reproduce this on Civtest with the same config, because when this first came up some testers were saying it was fine.

**Comment by ttk2** [comment]
should I sync up this config with the civtest one?

**Comment by WildWeazel** [comment]
do a diff between Civcraft and Civtest, out of curiosity

**Comment by drjawa** [comment]
<    #persistent_growth_period: 24
---
>    persistent_growth_period: 24

Only changes that came up aside from database stuff.

**Comment by staygroovy57** [comment]
@WildWeazel  Very sorry, been rather busy IRL and haven't had time to go on Civtest to find desert and TaigaHills biomes and try this out.

**Comment by ttk2** [comment]
no clue, how bad would it be to regen the map to try and find some? I guess I could map it.

**Comment by BlackXnt** [comment]
@ttk2 don't regen the map please it took us quite a bit of time to build all the factories.

**Comment by ttk2** [comment]
will do.

**Comment by gipsy-king** [comment]
While refactoring for persistent trees I found a bug where bukkit returns the wrong `TreeType` from saplings. This means that this issue is *probably* fixed, @staygroovy57 can you test again now?

**Comment by staygroovy57** [comment]
@gipsy-king While I was hoping to give you good news, I am sad to report that not a single tree grew in that area since I started AFK'ing it about 40 minutes ago. I do have another tree farm in a 'Taiga' biome and it doesn't suffer from this issue, only 'TaigaHills'

I believe that Realistic Biomes does not associate 'Taiga Hills' properly to what it should be, somehow.

**Comment by gipsy-king** [comment]
Spot on, looks like we have TAIGA and TAIGA_HILLS, but we're missing [TAIGA_MOUNTAINS](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/block/Biome.html#TAIGA_MOUNTAINS) in the config for `cold` and `cold_forest`.

**Comment by WildWeazel** [comment]
Aren't those new 1.8 biomes? Our map is from the 1.6 era.

**Comment by gipsy-king** [comment]
Mhh yea looks like 1.8. Could they still appear on our map from a chunk corruption/reset?

**Comment by staygroovy57** [comment]
@gipsy-king The problem appears to have started at the same time that the dark oak and acacia trees were fixed in RB config (Fall 2014), which is way before we moved to 1.8

**Comment by gipsy-king** [comment]
I found a `TaigaHills` biome in civtest, reports that redwood can grow. I'll report back after 3 hours.
![](http://i.imgur.com/RHQVz6r.png?1)

**Comment by WildWeazel** [comment]
>8 hours ago

@gipsy-king has died of hypothermia

**Comment by gipsy-king** [comment]
after unfreezing myself, the redwood tree has grown in `TaigaHills` on civtest.

**Comment by gipsy-king** [comment]
@staygroovy57 or anyone, got coords of a TaigaHills biome *on the main server*, to try to reproduce there since it doesn't reproduce on civtest?

**Comment by staygroovy57** [comment]
@gipsy-king The coords were in the screenshot I gave @WildWeazel 

They are -11,500 (78) -9258

**Comment by Goldmattress** [comment]
I checked, this appears to be fixed with the latest persistence update appears to have fixed it

**Comment by gipsy-king** [comment]
Now I'm getting infinite time on the civcraft.co test server (I think that one was only temporary for 1.8.7?) for redwood in TaigaHills. 

Closing since it seems to be fixed on main, reopen if necessary.

**Comment by ttk2** [comment]
the civtest server still shows that old issues where leaving RB alone for a long amount of time will cause it to lose DB connection, so that might be causing some strange behavior over there if you don't restart before doing much serious testing.

**Comment by ProgrammerDan** [comment]
That's no bueno, is there a keep-alive on the connection? I'll have to look into it, all the DB code needs gutting and replacing.

**Comment by gipsy-king** [comment]
That shouldn't affect getting the rate when hitting a block though

**Comment by ttk2** [comment]
@ProgrammerDan , @rourke750  plans to move all DB stuff into Civmod core and all the plugins just go through it for DB usage, should like a plan? 

@gipsy-king  probably not, its running the same config as prod though AFIK, try changing it out if its suspect.

**Comment by staygroovy57** [comment]
I just went to the second location I had reported which was a desert biome and acacias are finally growing there as well! So it would appear as if these problems have been eradicated. 

I would love to actually understand why the bug stopped acting up the way it was. 

Thanks to all concerned for staying on top of the issue. 

*Incidentally, it feels a bit odd that I may have been the only person to ever try growing acacia trees in the desert*

**Comment by ProgrammerDan** [comment]
@staygroovy57 Goodsprings outside MTA grows Acacia extensively in their desert, have for months. I used to buy them out quicker then they could harvest.

@ttk2 Tempered, but yes. I know some swank enterprise-level patterns for Generic DAO that I can embed in Civmod Core, including all the connection pooling and over icky stuff that new and old devs hate dealing with. Then each plugin could deal with Data Objects, instead of database calls (unless they really, really want to). They'd define their "data schema" simply as the primitives they are interested in and relationships between them, the DAO core patterns would do the rest. We'll see how it works out.

**Comment by ttk2** [comment]
@ProgrammerDan  cool stuff is cool, but changing the existing plugins that much is a big overhaul if it does not amount to large performance/simplicity improvements over just having them let someone else manage the connection and they just send queries.

**Comment by ProgrammerDan** [comment]
@ttk2 Reasonable. I'd definitely focus on the connection handling first. The DAO stuff is more than just cool, it's the "right way" to do things, but I get where you are coming from.

**Comment by staygroovy57** [comment]
@ProgrammerDan Well, I sure wish I could understand:
- why this issue happened concurrently with the RB fix to stop Dark Oak from growing everywhere?
- what it is that was changed that has allowed them start to grow again (TaigaHills) and for the first time (Desert)

Otherwise it may well be time to start giving credence to [e-lego voodo and unexplained phenomenons](https://en.wikipedia.org/wiki/Ghost_in_the_Shell). Minecraft religion and miracles can't be that far behind!

**Comment by WildWeazel** [comment]
@ProgrammerDan generic DAO support would be nice for Contraptions

**Comment by ProgrammerDan** [comment]
@WildWeazel Been thinking it'll be kind of a base requirement given the component architecture. I'll make my way to it soon :).

@staygroovy57  If I could answer all those I'd sell out to Microsoft and walk away with millions of bucks in my pockets

**Comment by gipsy-king** [comment]
@staygroovy57 all I know is that there has been a bug in spigot/bukkit where getting the tree species of a sapling was wrong. It would get the data value exactly like it does for logs, which have different block ids and different data value bitmask. When I added persistent trees I was going crazy with getting the wrong tree species, and once I confirmed the bug I just mapped the data values myself.

IIRC birch, jungle and generic tree species were not affected by the bug, they happened to match the data value for logs, but the others didn't.
